docs/source/cpp/dataset.rst

```diff
@@ -334,6 +334,25 @@ altogether if they do not match the filter:
    :linenos:
    :lineno-match:
 
+Partitioning performance considerations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Partitioning datasets can improve performance when reading datasets, but have several 
+potential costs when reading and writing:
+
+#. Can significantly increase the number of files to write. The number of partitions is a
+   floor for the number of files in a dataset. If you partition a dataset by date with a 
+   year of data, you will have at least 365 files. If you further partition by another
+   dimension with 1,000 unique values, you will have 365,000 files. This can make it slower
+   to write and increase the size of the overall dataset because each file has some fixed
+   overhead. For example, each file in parquet dataset contains the schema.
+#. Multiple partitioning columns can produce deeply nested folder structures which are slow
+   to navigate because they require many recusive "list directory" calls to discover files.
+   These operations may be particularly expensive if you are using an object store 
+   filesystem such as S3. One workaround is to combine multiple columns into one for
+   partitioning. For example, instead of a schema like /year/month/day/ use /YYYY-MM-DD/.
+ 
+
```

Comment:
Both of these are in the "cons" section.  It might be worth adding a bit more body to "can improve the performance when reading datasets".

There are two advantages (but really only one):

 * We need multiple files to read in parallel.
 * Smaller partitions allow for more selective queries.  E.g. we can load less data from the disk.

We should also mention (here or elsewhere) that everything that applies here for # of files also applies for # of record batchs (or # of row groups in parquet).  It's possible to have 1 file with way too many row groups and get similar performance issues.

Reply:
Probably a reasonable "rule of thumb" might be to structure files so that each column of data is at least 4MB large.  This is somewhat arbitrary when it comes to data/metadata ration but 4MB is also around the point where an HDD's sequential vs random reads tradeoff starts to fall off.  Although for bitmaps the requirement for 32 million rows can be a bit extreme / difficult to satisfy.

Reply:
I suppose compression complicates things too :shrug:

Reply:
Good point on those two advantages. I will add that to discussion.

Row group settings are (somewhat strangely) in the generic [FileSystemDatasetWriteOptions](https://github.com/apache/arrow/blob/331d3864fb524d276a86a31cd16cecbb67052bff/cpp/src/arrow/dataset/file_base.h#L385), so actually makes sense to discuss around here.

I feel somewhat reticent to specify too exact rules of thumb for file size. (1) that might change for Arrow over time as performance improvements are made. (2) that may be very different depending on use case (and compression as you point out); and (3) that may vary depending on who the reader is (it might be Spark or something else rather than Arrow C++.) What do you think about just pointing out cases that are pathological? For example, partitioning to file sizes less than a few MB means the overhead of the filesystem and the metadata outweighs any filtering speedups. And partitioning to file sizes of 2GB+ means not enough parallelism or OOM errors in many cases.

Reply:
@westonpace rewrote based on your feedback.

Reply:
Yes, the dataset writer has a lot of control now over row group sizing.  This is because a batch with 10,000 rows might arrive at the dataset writer and get partitioned into 100 batches with 100 rows based on the partitioning keys.  If those 100 row batches were delivered to the file writers then the file writers would write tiny batches.  Putting the "queue in memory until we have enough data" logic in the dataset writer instead of the file writers allowed us to solve that in one spot.

I agree on your thoughts for a specific limit and really like the idea of pointing out pathological cases.